joomla/joomla-cms
# [4.0] Add locked field to extensions table, prevent uninstalling core extensions, restructure protected extension list

Porting of PR #17219 to J4

All credits to @mbabker. Thank you!

Thanks @richard67 for updating the SQL files.

Text below copied from PR #17219.

 ### Summary of Changes
This is PR #13037. See the original PR for added context and past conversation.
 
The long and short is this adds a new field, `locked`, to the extensions table, and splits the current definition of the `protected` field. This also updates the protected extension list to make the list only include admin components, core libraries, and the files extension that Joomla tracks itself with (this means all modules and plugins are now unprotected and can be enabled/disabled consistently).
 
Current:
 
* `protected` indicates an extension which cannot be disabled or uninstalled
 
New:
 
* `protected` indicates an extension which cannot be disabled
 
* `locked` indicates an extension which cannot be uninstalled

 
 #### Why This Distinction Matters
Uninstalling core extensions can be problematic, and honestly it's not very effective given our current packaging and upgrading solutions. Even if you do uninstall the extensions, they end up back on your site's filesystem during the update process because we don't make the package extraction step aware of the database and inherently the installed extensions. Also, uninstalling components takes their tables with them, and if an update includes a SQL delta for one of those tables, it causes the update to fail over. So we should take an extra step to protect users from doing things that could be dangerous for their sites. Next, an uninstallable extension should not mean that the extension must be enabled. The only extensions which should be protected are those which if disabled would critically bring down the site, every other aspect of the extension listing should be controllable by the site administrator.

### Testing Instructions

It needs to test the same changes in 2 scenarios:
1. New installation with this PR applied.
2. Update from 3.10-dev or 3.10-alpha to J4 + this PR with a patched updated package.

Then, after the new installation or the update, it needs to test in both cases:
- The user should be unable to uninstall any core extension (on a correct install this is any extension with an ID less than 10000).
- The user should not be able to disable extensions which are critical to the management of the site (i.e. the extension manager, plugin manager, or update component, any of the libraries, or Joomla files extension) or are coupled to the core application stack (i.e. com_categories or com_content). All other extensions should be able to be disabled through the Extension Manager interface.

For scenario 1 new installation, apply the patch of this PR to a clean 4.0-dev branch when having a git clone, or use the patched full package download for this PR (see below).

For scenario 2, use the patched update package download for this PR (see below).

You can find a link to the downoads page when expanding the test results at the bottom of this PR on GitHub, if not already expanded:
![j4-pr-28462-link-to-packages-1](https://user-images.githubusercontent.com/7413183/77826474-fb225200-710f-11ea-813e-4a94d69718fb.png)

![j4-pr-28462-link-to-packages-2](https://user-images.githubusercontent.com/7413183/77826482-05dce700-7110-11ea-96f9-c4f651f0bd85.png)

The link leads to a download page where you can donwnload patched installation and update packages for this PR. By the commit number you can check if they fit to the last commit of this PR, i.e. they are up to date:
![j4-pr-28462-link-to-packages-3](https://user-images.githubusercontent.com/7413183/77831449-87903d00-712f-11ea-97fa-fcc730474d24.png)

There is also a custom update URL given for the update package so you can use the online update with that custom URL.

Btw. this is a standard feature for new or newly modified PR's on the 4.0-dev branch, so it can be used in the same way for testing other PR's, too.

### Documentation Changes Required
 We should probably add a page explaining the various extension states and what these columns exist for if one doesn't already exist.
